src/Container/LCParticleContainer.cpp

```diff
@@ -240,144 +308,150 @@ void LCParticleContainer::handleBoundaryAction() {
-  int x;
-  int y;
-  int z;
-  if(id.at(0) == -1) {
-    x = cell_count.at(0);
-  }else if(id.at(0) == cell_count.at(0)) {
-    x = -1;
-  } else {
-    x = cell_count.at(0)-id.at(0);
-  }
-  if(id.at(1) == -1) {
-    y = cell_count.at(1);
-  }else if(id.at(1) == cell_count.at(1)) {
-    x = -1;
-  } else {
-    y = cell_count.at(1)-id.at(1);
-  }
-  if(id.at(2) == -1) {
-    z = cell_count.at(2);
-  }else if(id.at(2) == cell_count.at(2)) {
-    z = -1;
-  } else {
-    z = cell_count.at(2)-id.at(2);
-  }
-  return {x, y, z};
+std::array<double, 3> LCParticleContainer::findOponentXYZ(Particle* p) {
+
+  //calculates Cell ID
+  const int x_id = floor(p->getX().at(0) / cell_size.at(0));
+  const int y_id = floor(p->getX().at(1) / cell_size.at(1));
+  const int z_id = floor(p->getX().at(2) / cell_size.at(2));
 
+  // finds oponentCell ID
+  std::array<int, 3> oponentCellID = findOponentCellID({x_id, y_id, z_id});
+
+  //calcs the rest
+  return {oponentCellID.at(0) * cell_size.at(0) + p->getX().at(0) - x_id * cell_size.at(0),
+            oponentCellID.at(1) * cell_size.at(1) + p->getX().at(1) - y_id * cell_size.at(1),
+            oponentCellID.at(2) * cell_size.at(2) + p->getX().at(2) - z_id * cell_size.at(2)};
 }
 
-std::array<double, 3> LCParticleContainer::findOponentXYZ(std::array<double, 3> XYZ) {
-  double x;
-  double y;
-  double z;
-  if(XYZ.at(0) == -1) {
-    x = cell_count.at(0) * cell_size.at(0) + XYZ.at(0);
-  }else if(XYZ.at(0) == cell_count.at(0)) {
-    x = -1 + XYZ.at(0);
-  } else {
-    x = cell_count.at(0) * cell_size.at(0) - XYZ.at(0);
+std::array<int, 3> LCParticleContainer::findOponentCellID(std::array<int, 3> ID) {
+  std::array<int, 3> return_id = ID;
+  if(ID.at(0) < 0) {
+    return_id.at(0) = cell_count.at(0) -1;
+  } else if(ID.at(0) >= cell_count.at(0)) {
+    return_id.at(0) = 1;
   }
-  if(XYZ.at(1) == -1) {
-    y = cell_count.at(1) * cell_size.at(1) + XYZ.at(1);
-  }else if(XYZ.at(1) == cell_count.at(1)) {
-    x = -1 + XYZ.at(1);
-  } else {
-    y = cell_count.at(1)*cell_size.at(1)-XYZ.at(1) ;
+  if(ID.at(1) < 0) {
+    return_id.at(1) = cell_count.at(1) -1;
+  } else if(ID.at(1) >= cell_count.at(1)) {
+    return_id.at(1) = 1;
   }
-  if(XYZ.at(2) == -1) {
-    z = cell_count.at(2) * cell_size.at(2) + XYZ.at(2);
-  }else if(XYZ.at(2) == cell_count.at(2)) {
-    z = -1 + XYZ.at(2);
-  } else {
-    z = cell_count.at(2)*cell_size.at(2)-XYZ.at(2);
+  if(ID.at(2) < 0) {
+    return_id.at(2) = cell_count.at(2) -1;
+  } else if(ID.at(2) >= cell_count.at(2)) {
+    return_id.at(2) = 1;
   }
-  return {x, y, z};
+  return return_id;
 }
 
 void LCParticleContainer::calcWithHalo(Particle *p, std::array<double, 3> x_arg, std::array<double, 3> v_arg) {
-  double m_arg = p->getM();
-  int type_arg = p->getType();
-  Particle haloParticle = Particle(x_arg, v_arg, m_arg, type_arg);
+  Particle haloParticle = Particle(x_arg, v_arg, p->getM(), p->getType());
```

Comment:
this is what I meant, you are only using velocity for creating haloParticle, which is not necessary. So, just remove all velocity calculations and you will save up on a lot of computation time